### PR TITLE
Added callback to customize attachment storage strategy.

### PR DESCRIPTION
**Pull request by Seb-C**
This adds a callback allowing to customize the default storage strategy for attachments.

Example (model) :

``` php
public static function _init()
{
    static::$_attachment['file'] = array(
        'attached_callback' => function($id) {
            return 'foo/bar/'.$id;
        },
    );
}
```

To access the attachment, you also need to alter the alias in the new 404 event :

``` php
\Event::register_function('404.attachmentMatch', function($params) {
    $params['alias'] = dirname(dirname($params['alias'])).'/';
});
```

### CODE REVIEW

File: framework/classes/attachment.php [hunk 1]
```diff
@@ -56,10 +56,10 @@ public static function forge($attached, $config)
     public function __construct($attached, $config)
     {
         if (!is_array($config)) {
-            $config = \Config::load($config, true);
+            $config = \Config::load($config, false);
```
Comment (jguyomard) [inline framework/classes/attachment.php]: Why this change?

**Comment by Seb-C** [comment]
@jguyomard It indeed wasn't necessary (I misunderstood the role of this argument). I fixed it.

File: framework/classes/attachment.php [hunk 2]
```diff
@@ -251,6 +256,13 @@ public function setId($id)
     {
         $path = $this->path();
         $this->attached = $id;
+
+        if(!empty($config['attached_callback']) && is_callable($config['attached_callback'])) {
+            $this->attached = $config['attached_callback']($id);
```
Comment (jguyomard) [inline framework/classes/attachment.php]: $config is undefined here

```
$this->config['attached_callback']($id, $this->config);
```

**Comment by jguyomard** [comment]
:+1:

File: framework/classes/attachment.php [hunk 3]
```diff
@@ -251,6 +256,13 @@ public function setId($id)
     {
         $path = $this->path();
         $this->attached = $id;
+
+        if(!empty($this->config['attached_callback']) && is_callable($this->config['attached_callback'])) {
```
Comment (smeeckaert) [inline framework/classes/attachment.php]: Missing space after if

File: framework/classes/attachment.php [hunk 4]
```diff
@@ -82,7 +82,12 @@ public function __construct($attached, $config)
         }
         $config['alias'] = rtrim(str_replace(DS, '/', $config['alias']), '/').'/';
 
-        $attached = preg_replace('`/`Uu', '_', $attached);
+        if(!empty($config['attached_callback']) && is_callable($config['attached_callback'])) {
```
Comment (smeeckaert) [inline framework/classes/attachment.php]: Missing space

**Comment by smeeckaert** [comment]
Little typos otherwise it seems good.

File: framework/classes/orm/attachment.php [hunk 5]
```diff
@@ -46,7 +46,7 @@ public function __construct($from, $name, array $config)
         $this->cascade_delete  = array_key_exists('cascade_delete', $config)
             ? $config['cascade_delete'] : $this->cascade_delete;
 
-        $attachment_config = array_diff($config, array('key_from' => '', 'cascade_save' => '', 'cascade_delete' => ''));
+        $attachment_config = array_diff_assoc($config, array('key_from' => '', 'cascade_save' => '', 'cascade_delete' => ''));
```
Comment (shaoshiva) [inline framework/classes/orm/attachment.php]: I think it should be neither _array_diff_ nor _array_diff_assoc_ but _array_diff_key_ or even better _\Arr::filter_keys()_ with third parameter to "true" : 
http://fuelphp.com/docs/classes/arr.html#/method_filter_keys

**Comment by smeeckaert** [comment]
:+1:
